### PR TITLE
NetCDF resources migration to composite resource type

### PR DESCRIPTION
**Pull request by pkdash**
### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [x] Automated Testing

- [x] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. Testing locally:
    - Deploy the branch '3135-netcdf-resources-migration'. 
    - Edit the 'hs_core/templates/pages/create-resource.html' file to enable creating NetCDF resource.
    - Create couple of NetCDF  resources with a nc file upload using the local hydroshare website. You can use nc file (netcdf_valid.nc) that is in the hs_app_netCDF/tests folder for upload. Verify that each of the NetCDF resources you created has one nc file and one txt file. 
    - Run the management command 'migrate_netcdf_resources'. Output of the command should indicate the number of netcdf resources that were converted to composite resources and that there are 0 number of netcdf resources after the conversion. 
    - Click 'MY  RESOURCES' link on the webpage. You should see all resources are of type composite resource.
    - Go to the landing page of one of the composite resources. You should see that it contains one netcdf aggregation. Verify the metadata of this aggregation. Verify that the aggregation contains the original nc and txt files.
    - Download the bag for one of the composite resources. Verify the contents of the bag file that the expected resources files are there. Check that the 'resourcemap.xml' and 'resourcemetadata.xml' files do not contain the word 'NetCDFResource'.
2. Testing with real NetCDF resources (**this test is essential for approving this PR**):
    - Deploy branch '3135-netcdf-resources-migration' to deployment environment similar to 'beta' deployment where it would have access to a copy of real hydroshare resources.
    - Make a note of 'title' and 'resource id' of few NetCDF resources so that you will able to check those resources after they are converted to composite resource.
    - Run the management command 'migrate_netcdf_resources'. Output of the command should indicate the number of netcdf resources that were converted to composite resources and that there are 0 number of netcdf resources after the conversion. 
   - Use the web interface to check that there are no netcdf resources. Use your note to verify that those resources have been converted to composite resource. Verify that such a converted composite resource has one netcdf aggregation (containing one nc file and one txt file)  and you can view the aggregation level metadata. Check that you are able to download the bag file for one of these converted resources. Verify the contents of the bag file.

### CODE REVIEW

**Comment by pkdash** [comment]
If possible would like a dry run of this new management command on a copy of www resources before I start writing code for migrating raster resources.

File: hs_core/management/utils.py [hunk 1]
```diff
@@ -28,6 +28,32 @@
...
+    :param orig_meta_obj: metadata object of the resource that is getting converted to
+    composite resource
+    :param comp_res: converted composite resource
+    :return:
+    """
+    single_meta_elements = ['title', 'type', 'language', 'rights', 'description',
+                            'publisher']
+    multiple_meta_elements = ['creators', 'contributors', 'coverages', 'subjects',
+                              'dates', 'formats', 'identifiers', 'sources', 'relations',
+                              'funding_agencies']
```
Comment (sblack-usu) [inline hs_core/management/utils.py]: 'modified' in dates isn't getting copied over (calls to last_updated after the conversion is failing).

File: hs_core/management/commands/migrate_netcdf_resources.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        logger = logging.getLogger(__name__)
+        resource_counter = 0
+        to_resource_type = 'CompositeResource'
+        msg = "THERE ARE CURRENTLY {} MULTIDIMENSIONAL RESOURCES PRIOR TO CONVERSION.".format(
+            NetcdfResource.objects.all().count())
+        logger.info(msg)
+        print(">> {}".format(msg))
+
+        for nc_res in NetcdfResource.objects.all():
+            # get the nc file name which needs to be used to create a new folder
```
Comment (sblack-usu) [inline hs_core/management/commands/migrate_netcdf_resources.py]: It'd be helpful to wrap each item with a try/except block with a message if it fails.  Right now, if something fails, it blocks the whole thing.

File: hs_core/management/utils.py (hunk 1 above)
Comment (pkdash) [reply hs_core/management/utils.py]: I am not able to reproduce this. I am seeing the last updated date/time on UI after the conversion. I also see resource modified date in resourcemetadata.xml file of the converted resource.

File: hs_core/management/commands/migrate_netcdf_resources.py (hunk 2 above)
Comment (pkdash) [reply hs_core/management/commands/migrate_netcdf_resources.py]: Have added error checking for potential errors.

File: hs_core/management/utils.py (hunk 1 above)
Comment (pkdash) [reply hs_core/management/utils.py]: I think it is working now as I updated this with develop today.

**Comment by pkdash** [comment]
@sblack-usu Can you take another look at this? Thanks.

**Comment by sblack-usu** [comment]
The dev-hs machines are having trouble right now and beta is busy with the release.  I'll run a test of this next week after the release on beta, but otherwise looks good.

**Comment by pkdash** [comment]
> The dev-hs machines are having trouble right now and beta is busy with the release. I'll run a test of this next week after the release on beta, but otherwise looks good.

Sounds good.

**Comment by sblack-usu** [comment]
I'm getting this error when running the migration on beta.
image

**Comment by sblack-usu** [comment]
```
File "/hydroshare/hs_core/management/commands/migrate_netcdf_resources.py", line 97, in handle
    comp_res.save()
```

**Comment by sblack-usu** [comment]
@pkdash - I had to make a couple of changes to get this to run on the latest develop.  I've run the migration on beta, please take a look and review my changes.

**Comment by pkdash** [comment]
@sblack-usu Your change looks fine to me. Should I merge then?

**Comment by sblack-usu** [comment]
yes, please

**Pull request merged**